Adopt std::partial_ordering now that we require C++20

cdumez opened the pull request:
#### 081bc346f6bfd314ec53525c8b9afd1a52ff3bae
<pre>
Adopt std::partial_ordering now that we require C++20
<a href="https://bugs.webkit.org/show_bug.cgi?id=256406">https://bugs.webkit.org/show_bug.cgi?id=256406</a>

Reviewed by Darin Adler.

* Source/WebCore/Modules/highlight/Highlight.cpp:
(WebCore::repaintRange):
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::characterOffsetsInOrder):
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::misspellingRange const):
(WebCore::AccessibilityObject::rangeOfStringClosestToRangeInDirection const):
* Source/WebCore/animation/WebAnimationUtilities.cpp:
(WebCore::compareDeclarativeAnimationOwningElementPositionsInDocumentTreeOrder):
* Source/WebCore/dom/BoundaryPoint.cpp:
(WebCore::order):
(WebCore::treeOrder):
(WebCore::treeOrderForTesting):
* Source/WebCore/dom/BoundaryPoint.h:
* Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
(WebCore::FragmentDirectiveRangeFinder::rangeOfStringInRange):
* Source/WebCore/dom/Node.cpp:
(WebCore::treeOrder):
(WebCore::treeOrderForTesting):
* Source/WebCore/dom/Node.h:
(WebCore::PartialOrdering::PartialOrdering): Deleted.
(WebCore::is_eq): Deleted.
(WebCore::is_lt): Deleted.
(WebCore::is_gt): Deleted.
(WebCore::is_neq): Deleted.
(WebCore::is_lteq): Deleted.
(WebCore::is_gteq): Deleted.
* Source/WebCore/dom/Position.cpp:
(WebCore::treeOrder):
(WebCore::documentOrder):
* Source/WebCore/dom/Position.h:
(WebCore::operator&lt;):
(WebCore::operator&gt;):
(WebCore::operator&gt;=):
(WebCore::operator&lt;=):
* Source/WebCore/dom/RadioButtonGroups.cpp:
(WebCore::RadioButtonGroup::members const):
* Source/WebCore/dom/Range.cpp:
(WebCore::Range::setStart):
(WebCore::Range::setEnd):
(WebCore::Range::comparePoint const):
(WebCore::Range::compareNode const):
(WebCore::Range::compareBoundaryPoints const):
* Source/WebCore/dom/SimpleRange.cpp:
(WebCore::contains):
(WebCore::treeOrder):
(WebCore::intersects):
(WebCore::compareByComposedTreeOrder):
* Source/WebCore/dom/SimpleRange.h:
* Source/WebCore/editing/ApplyStyleCommand.cpp:
(WebCore::ApplyStyleCommand::updateStartEnd):
(WebCore::ApplyStyleCommand::applyRelativeFontStyleChange):
(WebCore::ApplyStyleCommand::removeInlineStyle):
* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::simplifyMarkup):
* Source/WebCore/editing/TextIterator.cpp:
(WebCore::characterCount):
* Source/WebCore/editing/VisiblePosition.cpp:
(WebCore::documentOrder):
* Source/WebCore/editing/VisiblePosition.h:
(WebCore::operator&lt;):
(WebCore::operator&gt;):
(WebCore::operator&lt;=):
(WebCore::operator&gt;=):
* Source/WebCore/editing/VisibleSelection.cpp:
(WebCore::VisibleSelection::setBaseAndExtentToDeepEquivalents):
(WebCore::VisibleSelection::setWithoutValidation):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::findTextMatches):
(WebCore::replaceRanges):
* Source/WebCore/testing/Internals.cpp:
(WebCore::string):
* Tools/TestWebKitAPI/Tests/WebCore/DocumentOrder.cpp:
(TestWebKitAPI::string):
(TestWebKitAPI::operator-):
(TestWebKitAPI::allPositionTypeFailures):

Canonical link: <a href="https://commits.webkit.org/263767@main">https://commits.webkit.org/263767@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7ea74af40c35b5c282b37ddab0de7f6e95a28a11

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/5700 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/5858 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/6048 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/7252 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/6097 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/5702 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/6080 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/5829 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/7851 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/5806 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/5851 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/5116 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/7303 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/3328 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/5130 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/12879 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/5196 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/5207 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/5648 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/4624 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/5096 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1342 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/9210 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/5456 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/dom/BoundaryPoint.cpp:
```diff
@@ -64,40 +66,38 @@ static bool isOffsetBeforeChild(ContainerNode& container, unsigned offset, Node&
     return false;
 }
 
-// FIXME: Once we move to C++20, replace with the C++20 <=> operator.
```
Why are you removing this comment?

cdumez replied:
Not intentional. I can probably adopt <=> instead.

darinadler commented on Source/WebCore/dom/Node.h:
```diff
@@ -748,38 +749,21 @@ class Node : public EventTarget {
-constexpr bool is_neq(PartialOrdering);
-constexpr bool is_lteq(PartialOrdering);
-constexpr bool is_gteq(PartialOrdering);
+// FIXME: Se should be using the equivalent functions in the std namespace
```
Typo here: “Se”

darinadler commented on Source/WebCore/Modules/highlight/Highlight.cpp:
```diff
@@ -56,7 +56,7 @@ static void repaintRange(const SimpleRange& range)
     // FIXME: Unclear precisely why we need to handle out of order cases here, but not unordered cases.
     auto sortedRange = range;
-    if (is_gt(treeOrder<ComposedTree>(range.start, range.end)))
+    if (WebCore::is_gt(treeOrder<ComposedTree>(range.start, range.end)))
```
This code is in the `WebCore` namespace. Do we really need to add the `WebCore` prefix? Maybe you global replaced after trying `std`?

cdumez replied:
It didn’t build without WebCore::. The compiler said it was ambiguous because of the same function coming from std. no idea why we have access to the std namespace without an explicit std::

darinadler commented on Source/WebCore/dom/BoundaryPoint.cpp:
```diff
@@ -27,11 +27,13 @@
 #include "BoundaryPoint.h"
 #include "ContainerNode.h"
 
+#include <compare>
```
I don’t understand where we are adding this include, where we are not adding it, and why.

cdumez replied:
I can clean this up. At first, I thought the is_eq errors were caused by missing includes so I tried added the included in the files where it failed to build. But it didn’t help

darinadler commented on Source/WebCore/dom/Node.h:
```diff
@@ -748,38 +749,21 @@ class Node : public EventTarget {
-constexpr bool is_lteq(PartialOrdering);
-constexpr bool is_gteq(PartialOrdering);
+// FIXME: We should be using the equivalent functions in the std namespace
+// but it doesn't work yet on some of our bots (rdar://87314077).
```
I think you should consider using all the others from `<compare>`, except for `is_eq` and `is_neq`.

cdumez replied:
I was hesitating doing that. I'll make the change.

cdumez committed:
Adopt std::partial_ordering now that we require C++20
https://bugs.webkit.org/show_bug.cgi?id=256406

Reviewed by Darin Adler.

* Source/WebCore/Modules/highlight/Highlight.cpp:
(WebCore::repaintRange):
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::characterOffsetsInOrder):
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::misspellingRange const):
(WebCore::AccessibilityObject::rangeOfStringClosestToRangeInDirection const):
* Source/WebCore/animation/WebAnimationUtilities.cpp:
(WebCore::compareDeclarativeAnimationOwningElementPositionsInDocumentTreeOrder):
* Source/WebCore/dom/BoundaryPoint.cpp:
(WebCore::order):
(WebCore::treeOrder):
(WebCore::treeOrderForTesting):
* Source/WebCore/dom/BoundaryPoint.h:
* Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp:
(WebCore::FragmentDirectiveRangeFinder::rangeOfStringInRange):
* Source/WebCore/dom/Node.cpp:
(WebCore::treeOrder):
(WebCore::treeOrderForTesting):
* Source/WebCore/dom/Node.h:
(WebCore::PartialOrdering::PartialOrdering): Deleted.
(WebCore::is_eq): Deleted.
(WebCore::is_lt): Deleted.
(WebCore::is_gt): Deleted.
(WebCore::is_neq): Deleted.
(WebCore::is_lteq): Deleted.
(WebCore::is_gteq): Deleted.
* Source/WebCore/dom/Position.cpp:
(WebCore::treeOrder):
(WebCore::documentOrder):
* Source/WebCore/dom/Position.h:
(WebCore::operator<):
(WebCore::operator>):
(WebCore::operator>=):
(WebCore::operator<=):
* Source/WebCore/dom/RadioButtonGroups.cpp:
(WebCore::RadioButtonGroup::members const):
* Source/WebCore/dom/Range.cpp:
(WebCore::Range::setStart):
(WebCore::Range::setEnd):
(WebCore::Range::comparePoint const):
(WebCore::Range::compareNode const):
(WebCore::Range::compareBoundaryPoints const):
* Source/WebCore/dom/SimpleRange.cpp:
(WebCore::contains):
(WebCore::treeOrder):
(WebCore::intersects):
(WebCore::compareByComposedTreeOrder):
* Source/WebCore/dom/SimpleRange.h:
* Source/WebCore/editing/ApplyStyleCommand.cpp:
(WebCore::ApplyStyleCommand::updateStartEnd):
(WebCore::ApplyStyleCommand::applyRelativeFontStyleChange):
(WebCore::ApplyStyleCommand::removeInlineStyle):
* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::simplifyMarkup):
* Source/WebCore/editing/TextIterator.cpp:
(WebCore::characterCount):
* Source/WebCore/editing/VisiblePosition.cpp:
(WebCore::documentOrder):
* Source/WebCore/editing/VisiblePosition.h:
(WebCore::operator<):
(WebCore::operator>):
(WebCore::operator<=):
(WebCore::operator>=):
* Source/WebCore/editing/VisibleSelection.cpp:
(WebCore::VisibleSelection::setBaseAndExtentToDeepEquivalents):
(WebCore::VisibleSelection::setWithoutValidation):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::findTextMatches):
(WebCore::replaceRanges):
* Source/WebCore/testing/Internals.cpp:
(WebCore::string):
* Tools/TestWebKitAPI/Tests/WebCore/DocumentOrder.cpp:
(TestWebKitAPI::string):
(TestWebKitAPI::operator-):
(TestWebKitAPI::allPositionTypeFailures):

Canonical link: https://commits.webkit.org/263767@main
